diffrax/_solver/srk.py

```diff
@@ -0,0 +1,649 @@
+    b: Union[FloatScalarLike, Float[np.ndarray, " s"]]
+    # b_error: Optional[Union[FloatScalarLike, Float[np.ndarray, " s"]]]
+
+    @abc.abstractmethod
+    def check(self) -> int:
+        ...
+
+
+@dataclass(frozen=True)
+class AdditiveNoiseCoefficients(AbstractNoiseCoefficients):
+    """
+    Class representing the noise coefficients for additive noise.
+
+    **Arguments:**
+
+    - `b`: The coefficient for the Brownian increment when computing the output.
+    - `a`: The coefficients in front of the Brownian increment at each stage.
+    - `b_error`: The coefficient for the Brownian increment when computing the
+        error estimate.
+    """
+
+    b: FloatScalarLike
+    # assuming SDE has additive noise we only need a 1-dimensional array
+    # of length s for the coefficients in front of the Brownian increment
+    # and/or Lévy areas (where s is the number of stages of the solver).
+    # This is the equivalent of the matrix a for the Brownian motion and
+    # its Lévy areas.
+    a: Float[np.ndarray, " s"]
+    # b_error: Optional[FloatScalarLike] = None
+
+    def check(self):
+        assert self.a.ndim == 1
+        return self.a.shape[0]  # return the number of stages
+
+
+@dataclass(frozen=True)
+class GeneralNoiseCoefficients(AbstractNoiseCoefficients):
+    """Class representing the noise coefficients for general (non-additive) noise.
+
+    **Arguments:**
+
+    - `b`: The coefficient for the Brownian increment when computing the output.
+    - `a`: The coefficients in front of the Brownian increment at each stage.
+    - `b_error`: The coefficient for the Brownian increment when computing the
+        error estimate.
+    """
+
+    b: Float[np.ndarray, " s"]
+    # If the SDE has non-additive noise, we need an equivalent of the
+    # matrix a, one for the Brownian motion and one for each type of
+    # Lévy area.
+    a: tuple[np.ndarray, ...]
+    b_error: Optional[Float[np.ndarray, " s"]] = None
+
+    def check(self):
+        assert self.b.ndim == 1
+        assert all((i + 1,) == a_i.shape for i, a_i in enumerate(self.a))
+        if self.b_error is not None:
+            assert self.b_error.ndim == 1
+            assert self.b_error.shape == self.b.shape
+        assert self.b.shape[0] == len(self.a) + 1
+        return self.b.shape[0]  # return the number of stages
+
+
+@dataclass(frozen=True)
+class StochasticButcherTableau:
+    """A Butcher Tableau for Stochastic Runge-Kutta methods."""
+
+    # Only supports explicit SRK so far
+    c: np.ndarray
+    b_sol: np.ndarray
+    b_error: Optional[np.ndarray]
+    a: list[np.ndarray]
+
+    # Coefficients for the Brownian increment
+    cfs_w: AbstractNoiseCoefficients
+    # Optional coefficients for the space-time Lévy area
+    cfs_hh: Optional[AbstractNoiseCoefficients] = None
+    # Optional coefficients for the space-time-time Lévy area
+    cfs_kk: Optional[AbstractNoiseCoefficients] = None
+
+    additive_noise: bool = True
```

Comment:
Can these be combined so that we have a single `AbstractNoiseCoefficients` (with three times as many fields), and then `additive_noise` is automatically determined by `isinstance(noise_coeffs, AdditiveNoiseCoefficients)`.

Basically I'm trying to more strongly follow the principle we're already doing here, of "making illegal types unrepresentable". Or put another way, it simpilfies the API for `StochasticButcherTableau`: one no longer has to worry about passing a mix of `AdditiveNoiseCoefficients` and `GeneralNoiseCoefficients`, nor about mismatching them with the value of `additive_noise`. The user's intent is communicated through their single choice of a class for the coefficients.

Reply:
I understand what you mean. Of course there will always be illegal representable stuff (mostly to do with b_error), but I'll do my best to exclude as many illegal possibilities.